### PR TITLE
libopaevfio: add opae_vfio_buffer_allocate_ex()

### PR DESCRIPTION
**Pull request by tswhison**
Adds parameter flags to the end of the call. Currently only
OPAE_VFIO_BUF_PREALLOCATED is supported.

Signed-off-by: Tim Whisonant <tim.whisonant@intel.com>

### CODE REVIEW

File: libopaevfio/opaevfio.c [hunk 1]
```diff
@@ -774,6 +780,124 @@ int opae_vfio_buffer_allocate(struct opae_vfio *v,
...
+
+	if (opae_vfio_iova_reserve(v, size, &ioaddr)) {
+		pthread_mutex_unlock(&v->lock);
+		return 4;
+	}
+
+	if (!(flags & OPAE_VFIO_BUF_PREALLOCATED)) {
+
+		if (*size > (2 * 1024 * 1024))
+			vaddr = mmap(ADDR, *size, PROT_READ|PROT_WRITE,
```
Comment (anandaravuri) [inline libopaevfio/opaevfio.c:825]: define macro for 2 * 1024 * 1024
define macro 4096

File: libopaevfio/opaevfio.c [hunk 2]
```diff
@@ -774,6 +780,124 @@ int opae_vfio_buffer_allocate(struct opae_vfio *v,
...
+		mem_alloc_put(&v->iova_alloc, ioaddr);
+		res = 7;
+		goto out_munmap;
+	}
+
+	node = opae_vfio_create_buffer(vaddr, *size, ioaddr, flags);
+	if (!node) {
+		ERR("malloc failed\n");
+		mem_alloc_put(&v->iova_alloc, ioaddr);
+		res = 8;
```
Comment (anandaravuri) [inline libopaevfio/opaevfio.c:870]: res = 8;  res = 7;, res = 6;   errors mismatch  with FPGA_RESULT  ENUM in OPAE API

Comment (tswhison) [reply libopaevfio/opaevfio.c:870]: I'm not using fpga_result.

File: libopaevfio/opaevfio.c (hunk 1 above)
Comment (tswhison) [reply libopaevfio/opaevfio.c:825]: why?

File: libopaevfio/opaevfio.c [hunk 3]
```diff
@@ -774,6 +780,124 @@ int opae_vfio_buffer_allocate(struct opae_vfio *v,
...
+	}
+
+	if (opae_vfio_iova_reserve(v, size, &ioaddr)) {
+		pthread_mutex_unlock(&v->lock);
+		return 4;
+	}
+
+	if (!(flags & OPAE_VFIO_BUF_PREALLOCATED)) {
+
+		if (*size > (2 * 1024 * 1024))
```
Comment (r-rojo) [inline libopaevfio/opaevfio.c:824]: If we're duplicating code from lines 824-839 then perhaps we should think about refactoring this chunk into its own private function.

**Review by r-rojo — APPROVED** [review]

**Pull request merged**